### ISSUE TITLE
Should Roslyn or SDK provide a mechanism to enable/disable some/all source generators?

### ISSUE THREAD

**Issue by ericstj** [dotnet/sdk#19501]
User's might need to disable a source generator for some reason (like a performance or compatibility issue).  In general we should fix those and it shouldn't be a very common occurrence, however we may want to have a mechanism available as a mitigation.

In writing up the docs for https://github.com/dotnet/docs/pull/24896#discussion_r661102362 we came to the conclusion that allowing users to toggle all framework source generators/analyzers wasn't particularly useful, and might even be dangerous.

https://github.com/dotnet/sdk/issues/1212 means that analyzers/generators from packages can't easily be excluded either.

Should we have targets in the SDK/MSBuild that formalize a file-based exclusion mechanism?  Should roslyn support some exclusion mechanism itself (through compiler parameters, editorconfig, etc)?

cc @jaredpar @chsienki @terrajobst @KathleenDollard @jmarolf

**Comment by ericstj** [comment]
@sharwell  pointed me to how Razor does this https://github.com/dotnet/sdk/blob/2fea637fd7dcc56a8d29329dd0c28de4309bd5f4/src/RazorSdk/SourceGenerators/RazorSourceGenerator.RazorProviders.cs#L17
cc @captainsafia 

That's pretty cool that we could use .editorconfig to make generator-specific settings.  Perhaps one thing we could do here is make all generators read this and and define their own setting to enable/disable.

**Comment by captainsafia** [comment]
Yep -- we use the flag to disable the source generator under different scenarios, like in VS 2019 where the incremental source generator API is not supported.

Speaking of incremental source generators, there's still some work pending their to make it easier to disable a generator in the runtime. What we currently implement is a bit of a "hack" on the generator side.

cc: @chsienki

**Comment by ericstj** [comment]
One thought is that we could have a property or item like "DisabledSourceGenerators" and pass that in via https://github.com/dotnet/roslyn/blob/2946c619d91c7279d69d71bbb521ea4647e29fc1/docs/features/source-generators.cookbook.md#consume-msbuild-properties-and-metadata then have all generators honor that in their Initialize methods.  Seems like it'd be better if roslyn provided something, that way we wouldn't rely on folks all having this common code.
